### PR TITLE
Model Viewer - Add residual table

### PR DESCRIPTION
**Pull request by eslickj**
## Summary/Motivation:

The main difference here is that I added a residual table, but it also includes some minor features, cleanup and reorganization.

## Changes proposed in this PR:
- Moved the residual calculation from the tree data model items to the central UIData object.  This way when I calculate constraints and expressions I can use the values in other views.  At this point it was motivated by the residual table view.
- Added a table of constraint residuals sorted by residual value
- To trees, added font color to rows to distinguish between blocks and the items of interest
- Added a model selector to select which pyomo model to view from models in the __main__ space.
- Added a menu item to set the working directory
- Minor renaming and reorganization.  I renamed pyqt_4or5 to just qt.  I renamed UISetup to UIData and split the class so UIData inherits form QObject and UIDataNoUI.  The no ui part is separate to enable as much testing as possible when qt is not present.  The only thing the UI part adds is emitting signals. 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/viewer/ui.py [hunk 1]
```diff
@@ -162,21 +116,55 @@ def __init__(self, *args, **kwargs):
...
+        """
+        if wdir is None:
+            # Show a dialog box for user to selet working directory
+            wd = QFileDialog(self, 'Working Directory', os.getcwd())
+            wd.setFileMode(QFileDialog.DirectoryOnly)
+        if wd.exec_() == QFileDialog.Accepted:
+            wdir = wd.selectedFiles()[0]
+        else:
+            wdir = None
+        # Change directoy if one was selected
```
Comment (blnicho) [inline pyomo/contrib/viewer/ui.py]: Typo: "directoy"

File: pyomo/contrib/viewer/ui.py [hunk 2]
```diff
@@ -162,21 +116,55 @@ def __init__(self, *args, **kwargs):
...
+
+        Args:
+            wdir (str): if None show a dialog to select, otherwise try to
+                change to this path
+            checked (bool): the triggered signal sends this, but it is not used
+        Returns:
+            (str): new working directory path
+        """
+        if wdir is None:
+            # Show a dialog box for user to selet working directory
```
Comment (blnicho) [inline pyomo/contrib/viewer/ui.py]: Typo: "selet"

File: pyomo/contrib/viewer/ui_data.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+import logging
+from pyomo.contrib.viewer.qt import *
+from pyomo.kernel import ComponentMap
+import pyomo.environ as pe
+
+_log = logging.getLogger(__name__)
+
+class UIDataNoUi(object):
+    """
+    This is the UIData object minus the singnals.  This is the base class for
```
Comment (blnicho) [inline pyomo/contrib/viewer/ui_data.py]: Typo: "singnals"

File: pyomo/contrib/viewer/ui_data.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        """
+        Lets the model setup be changed without emitting the updated signal
+        until the end_update function is called.
+        """
+        self._begin_update = True
+
+    def end_update(self, emit=True):
+        """
+        Sets the begin update flag to false.  Needs to be overloaded to also
+        emmit an update signal in the full UIData class
```
Comment (blnicho) [inline pyomo/contrib/viewer/ui_data.py]: Typo: "emmit"

File: pyomo/contrib/viewer/ui_data.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        """
+        self._begin_update = False
+
+    def emit_update(self):
+        """
+        Don't forget the overloaded this, passing instead of NotImplementedError
```
Comment (blnicho) [inline pyomo/contrib/viewer/ui_data.py]: There are a few typos in this docstring

File: pyomo/contrib/viewer/ui_data.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        so testing can co smoothly without Qt
+        """
+        pass
+
+    def emit_exec_refresh(self):
+        """
+        Don't forget the overloaded this, passing instead of NotImplementedError
```
Comment (blnicho) [inline pyomo/contrib/viewer/ui_data.py]: There are a few typos in this docstring

**Comment by eslickj** [comment]
Okay I think I got it all fixed.  I also tried to bump the test coverage a bit where it was easy to do.

**Comment by eslickj** [comment]
Should be good now.

**Pull request merged**